### DISCUSSION TITLE
How to generate `\cite{}` from Quarto citations?

### DISCUSSION THREAD

**Discussion by ramiromagno** [quarto-dev/quarto-cli#6233] [Q&A] [answered]
### Description

Because of https://github.com/ramiromagno/wiley-njd/blob/885a1827617952d3d01156ab68e6af8c8e118337/_extensions/wiley-njd/wiley-njd-v5/NJDnatbib.sty#L62-L63

I need to find a way of quarto citations always generating `\cite{}` in latex, and not `\citep{}` or `\citet{}`.

The two modes are meant to be changed only at the document class level.

**Comment by dragonstyle** [comment]
It looks like Pandoc doesn't support bibtex style cites at all (only biblatex). The suggestion I found was to redefine `citep`, `citet` etc as `cite`.

I'm not sure that this would work for your case, though. If not, an alternative would be to write a filter to actually render the `cites` Rough example from our asciidoc support:

```lua
return {
    Cite = function(el) 
      -- If quarto is going to be processing the cites, go ahead and convert
      -- them to a native cite
      if param(kAsciidocNativeCites) then
        local citesStr = table.concat(el.citations:map(function (cite) 
          return '<<' .. cite.id .. '>>'
        end))
        return pandoc.RawInline("asciidoc", citesStr);
      end
    end
}
```

**Reply by ramiromagno** [reply]
Thanks for your answer! But I have no idea how to adapt your code to my example though :)

**Reply by dragonstyle** [reply]
Here's an example that should be working - it just writes all cites as `\cite{}` without any regarding to the citation style. You could make this smarter, or controlled by an option, etc... with additional work.


**cite-filter.lua**
```lua
return {
    Cite = function(el) 
        if quarto.doc.is_format("pdf") then
            local citesStr = table.concat(el.citations:map(function (cite) 
                return '\\cite{' .. cite.id .. '}'
            end))
            return pandoc.RawInline("latex", citesStr);    
        end
    end
}
```

**test.qmd**
```markdown
---
title: Text document
authors:
  - name: Norah Jones
    affiliation: The University
    roles: writing
    corresponding: true
bibliography: references.bib
filters: 
  - cite-filter.lua
---

## Section
This is a simple placeholder for the manuscript's main document [@knuth84].
```

HTH!

**Reply by ramiromagno** [reply]
It does help a lot, thank you so much!

Could you be so kind as to point in the right direction to make it work for multiple citations: `[@Burton2013; @Berndt2011;@Kucharik2012]` -> `\cite{Burton2013, Berndt2011, Kucharik2012}`; currently, it generates `\cite{Burton2013}\cite{Berndt2011}\cite{Kucharik2012}`.

**Reply by cscheid** [reply]
You'll need to adapt this:

```
            local citesStr = table.concat(el.citations:map(function (cite) 
                return '\\cite{' .. cite.id .. '}'
            end))
```

Unfortunately, extending quarto for advanced use cases does require becoming familiar with Lua.

**Reply by ramiromagno** [reply]
I appreciate your help thus far! I will take it upon myself to learn Lua and ensure that I can start addressing these issues autonomously. Thank you both once more!

**Comment by ramiromagno** [comment, marked as answer]
For others that might be looking for the solution, this is what I came up with (use at your own discretion):

```lua
return {
  Cite = function(el)
    if quarto.doc.is_format("pdf") then
      local cites = el.citations:map(function(cite)
        return cite.id
      end)
      local citesStr = "\\cite{" .. table.concat(cites, ", ") .. "}"
      return pandoc.RawInline("latex", citesStr)
    end
  end
}
```

**Reply by nikitoshina** [reply]
Thank you for your snippet. I have made some improvements to identify the citations from a .bib file and handle references using \ref. (One could probably use Meta to extract the bibliography field.) The code works by extracting all the keys from the references.bib file using a simple regex. It then compares each entry in the Cite table with these keys. If a reference is found, all entries are wrapped in \cite; if not, they are wrapped in \ref. Note that it doesn’t work well if both references and citations are in the same block. However, it is useful when you have multiple citations and only some of them have references, as it will still wrap in \cite.

I hope this helps anyone encountering the same issue!

```
-- Purpose: Extract citation keys from a .bib file and use them to determine
-- whether to use \cite or \ref in the output.
local bibfile = "references.bib" -- Your bib file

function log(msg)
	quarto.log.output(msg)
end

local function anyTrue(t)
	for _, v in pairs(t) do
		if v == true then
			return true
		end
	end

	return false
end

log("running extract-citations")

-- Function to extract citation keys from a .bib file
function extractCitationKeys(bibfile)
	local citationKeys = {}
	for line in io.lines(bibfile) do
		local key = line:match("@%w+{([^,]+),")
		if key then
			citationKeys[key] = true
		end
	end
	return citationKeys
end

local citationKeys = extractCitationKeys(bibfile)

-- log(table.concat(citationKeys, ", "))

-- Function to process Cite elements
return {
	Cite = function(el)
		if quarto.doc.is_format("pdf") then
			local cites = el.citations:map(function(cite)
				return cite.id
			end)
			-- check whether all of them are in the bib file
			local pass = cites:map(function(cite)
				local out = citationKeys[cite] == true
				return out
			end)

			local all_cite = anyTrue(pass)

			log(all_cite)
			-- log(cites)

			local citesStr = ""

			if all_cite then
				citesStr = "\\cite{" .. table.concat(cites, ", ") .. "}"
			else
				citesStr = "\\ref{" .. table.concat(cites, ", ") .. "}"
			end

			return pandoc.RawInline("latex", citesStr)
		end
	end,
}
```
